src/Framework/AspNetCoreAnalyzers/src/Analyzers/RouteHandlers/DetectInvalidCustomBindingFormat.cs

```diff
@@ -0,0 +1,188 @@
+        {
+            if (!parameter.Type.IsReferenceType)
+            {
+                continue;
+            }
+
+            DetectInvalidMethodFormatFromParameter(
+                context: context,
+                parameterSymbol: parameter,
+                methodName: "BindAsync",
+                returnType: wellKnownTypes.ValueTaskOfT,
+                parametersValidator: (OperationAnalysisContext context, IMethodSymbol method) => ValidateBindAsyncParameters(method, wellKnownTypes),
+                modifiersValidator: ValidateCustomBindingModifiers,
+                onInvalidMessage: DiagnosticDescriptors.CustomBindingBindAsyncMustHaveAValidFormat);
+
+            DetectInvalidMethodFormatFromParameter(
+                context: context,
+                parameterSymbol: parameter,
+                methodName: "TryParse",
+                returnType: context.Compilation.GetSpecialType(SpecialType.System_Boolean),
+                parametersValidator: (OperationAnalysisContext context, IMethodSymbol method) => ValidateTryParseParameters(method, wellKnownTypes),
+                modifiersValidator: ValidateCustomBindingModifiers,
+                onInvalidMessage: DiagnosticDescriptors.CustomBindingTryParseMustHaveAValidFormat);
+        }
+    }
+
+    private static bool ValidateTryParseParameters(IMethodSymbol method, WellKnownTypes wellKnownTypes)
+    {
+        if (method.Parameters.Length < 2 || method.Parameters.Length > 3)
+        {
+            return false;
+        }
+
+        var isFirstParameterString = method.Parameters.First().Type.SpecialType == SpecialType.System_String;
+        var hasLastParameterOutModifier = method.Parameters.Last().RefKind == RefKind.Out;
+        var isIFormatProviderParameterValidIfExist = method.Parameters.Length < 3 || wellKnownTypes.IFormatProvider.IsAssignableFrom(method.Parameters[1].Type);
+
+        if (!isFirstParameterString || !hasLastParameterOutModifier || !isIFormatProviderParameterValidIfExist)
+        {
+            return false;
+        }
+
+        return true;
+    }
+
+    private static bool ValidateBindAsyncParameters(IMethodSymbol method, WellKnownTypes wellKnownTypes)
+    {
+        if (method.Parameters.Length < 1 || method.Parameters.Length > 2)
+        {
+            return false;
+        }
+
+        var isFirstParameterHttpContext = wellKnownTypes.HttpContext.IsAssignableFrom(method.Parameters[0].Type);
+        var isSecondParameterParameterInfoIfExists = method.Parameters.Length < 2 || wellKnownTypes.ParameterInfo.IsAssignableFrom(method.Parameters[1].Type);
+        if (!isFirstParameterHttpContext || !isSecondParameterParameterInfoIfExists)
+        {
+            return false;
+        }
+
+        return true;
+    }
+
+    private static bool ValidateCustomBindingModifiers(OperationAnalysisContext context, IMethodSymbol method)
+    {
+        if (method.DeclaredAccessibility != Accessibility.Public)
```

Comment:
Not sure we should be warning for non-public methods. We don't throw for non-public methods.